core/arch/arm/kernel/misc.c

```diff
@@ -0,0 +1,242 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/mobj.h>
+#include <optee_msg.h>
+#include <sm/optee_smc.h>
+#include <sm/sm.h>
+#include <trace.h>
+#include <util.h>
+
+/**
+ * extract_pages_from_params() - extract list of pages from
+ * OPTEE_MSG_ATTR_FRAGMENT parameters
+ *
+ * @params:	pointer to parameters array
+ * @pages:	output array of page addresses
+ * @num_params: number of parameters in array
+ *
+ * return:
+ *   page count on success
+ *   <0 on error
+ */
+int extract_pages_from_params(struct optee_msg_param *params, paddr_t *pages,
+			      uint32_t num_params)
+{
+	uint32_t pages_cnt = 0;
+
+	if (params[num_params-1].attr & OPTEE_MSG_ATTR_FRAGMENT)
+		return -1;
+
+	for (uint32_t i = 0; i < num_params; i++) {
+		uint32_t attr = params[i].attr & OPTEE_MSG_ATTR_TYPE_MASK;
+
+		switch (attr) {
+		case OPTEE_MSG_ATTR_TYPE_NEXT_FRAGMENT:
+			continue;
+		case OPTEE_MSG_ATTR_TYPE_TMEM_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT:
+			if (pages_cnt >= num_params)
+				return -1;
+			if (!(params[i].attr & OPTEE_MSG_ATTR_FRAGMENT) &&
+			    (i != num_params - 1))
+				return -1;
+			pages[pages_cnt++] = params[i].u.tmem.buf_ptr
+				& ~SMALL_PAGE_MASK;
+			break;
+		default:
+			return -1;
+		}
+	}
+	return pages_cnt;
+}
+
+/**
+ * map_params_buffer() - map parameters buffer into OP-TEE VA space
+ * @pa_params - physical pointer to parameters
+ * @num_params - number of parameters
+ *
+ * return:
+ *  struct shmem_mapping of mapped buffer on success
+ *  NULL on error.
+ */
+struct mobj *map_params_buffer(paddr_t pa_params, uint32_t num_params)
+{
+	struct mobj *mobj = NULL;
+	struct optee_msg_param *params;
+	paddr_t *pages = NULL;
+	uint32_t max_pages, pages_cnt;
+	size_t args_size;
+
+	args_size = OPTEE_MSG_GET_ARG_SIZE(num_params);
+	max_pages = args_size/SMALL_PAGE_SIZE + 1;
+
+	pages = calloc(max_pages, sizeof(paddr_t));
+	if (!pages)
+		goto err;
+
+	pages[0] = pa_params & ~SMALL_PAGE_MASK;
+	mobj = mobj_optee_shm_alloc(pages, 1, 0, 0);
+	if (!mobj)
+		goto err;
+
+	params = (struct optee_msg_param *)
+		mobj_get_va(mobj, pa_params & SMALL_PAGE_MASK);
+	pages_cnt = 1;
+	for (uint32_t i = 0; i < num_params; i++) {
+		if ((params->attr & OPTEE_MSG_ATTR_TYPE_MASK) ==
+		    OPTEE_MSG_ATTR_TYPE_NEXT_FRAGMENT) {
+			if (pages_cnt >= max_pages)
+				goto err;
+
+			pages[pages_cnt] = params->u.tmem.buf_ptr;
+			mobj_free(mobj);
+
+			mobj = mobj_optee_shm_alloc(pages + pages_cnt, 1, 0, 0);
+			if (!mobj)
+				goto err;
+
+			pages_cnt++;
+			params = (struct optee_msg_param *)mobj_get_va(mobj, 0);
+		} else {
+			params++;
+			if (((vaddr_t)params & SMALL_PAGE_MASK) == 0)
```

Comment:
It wouldn't hurt with a comment stating that we're checking that we're not overflowing into the next page.

Preferred style is `if (!((vaddr_t)params & SMALL_PAGE_MASK))`